Project export with 3d tasks

ActiveChooN opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

Added ability to export project dataset with 3d tasks
Depends on #3365 

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

Manually

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
~~- [ ] I have added tests to cover my changes~~
~~- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))~~
~~- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))~~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -797,20 +799,25 @@ def categories(self) -> dict:
 
     @staticmethod
-    def _load_categories(labels: list):
+    def _load_categories(meta: dict, dimension):
```
Can we make it another function? There is no reason to include it here.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -797,20 +799,25 @@ def categories(self) -> dict:
-        for _, label in labels:
+        user_info = {}
+        if dimension == DimensionType.DIM_3D:
+            user_info = {"name": meta['owner']['username'],
```
Can we make it another function? There is no reason to include it here.

ActiveChooN replied:
Fixed

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -952,12 +938,27 @@ def __init__(self, project_data: ProjectData, include_images: bool = False):
+                            task.data.get_upload_dirname(), kwargs['path'],
+                        )
+                        related_images = []
+                        image = Img.objects.get(id=i)
```
Can we use `fetch_related`?

ActiveChooN replied:
Sure

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -966,30 +967,48 @@ def _make_image(i, **kwargs):
             }
-            if include_images:
+            if self._dimension == DimensionType.DIM_3D:
+                dm_image = image_maker_per_task[frame_data.task_id](frame_data.id, **image_args)
+            elif include_images:
                 dm_image = image_maker_per_task[frame_data.task_id](frame_data.idx, **image_args)
```
Why one line uses `frame_data.id` and another one `frame_data.idx`?

ActiveChooN replied:
Because `make_image` functions are different depends on task type and dimension, and implementation of `make_image` function comes from some of 3D feature milestones.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -513,7 +513,7 @@ class ProjectData(InstanceLabelData):
     Tag = NamedTuple('Tag', [('frame', int), ('label', str), ('attributes', List[InstanceLabelData.Attribute]), ('source', str), ('group', int), ('task_id', int)])
     Tag.__new__.__defaults__ = (0, )
-    Frame = NamedTuple('Frame', [('task_id', int), ('subset', str), ('idx', int), ('frame', int), ('name', str), ('width', int), ('height', int), ('labeled_shapes', List[Union[LabeledShape, TrackedShape]]), ('tags', List[Tag])])
+    Frame = NamedTuple('Frame', [('task_id', int), ('subset', str), ('idx', int), ('id', int), ('frame', int), ('name', str), ('width', int), ('height', int), ('labeled_shapes', List[Union[LabeledShape, TrackedShape]]), ('tags', List[Tag])])
```
Can we reuse `TaskData` somehow? The code is almost the same. Maybe, extract a base class?

ActiveChooN replied:
Will do it in separate PR with refactoring.

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -940,9 +924,11 @@ def map_label(name): return label_cat.find(name)[0]
         return convert_cvat_anno_to_dm(cvat_frame_anno, label_attrs, map_label, self._format_type, self._dimension)
 
 class CVATProjectDataExtractor(datumaro.Extractor, CVATDataExtractorMixin):
```
Think of implementing this class as (roughly):

```python
task_datasets = { task_id: make_task_dataset(task) for task_id, task in tasks }
# append name mangling and other transforms like
for task in task_datasets:
  task.transform('rename')

dataset = Dataset.from_extractors(*task_datasets)
```

bsekachev commented:
I am not expert in `` cvat/apps/dataset_manager``, so I only tried this patch and was able to export both tasks and projects in 3D format. Just one question, if we dump 3D tasks we can **only choice between PCD and kitty formats**. But when we dump a project with one 3D tasks **it shows all the formats** (at the same time I cannot create a project with both 3D and 2D data). Is this expected behaviour?

bsekachev commented:
@zhiltsov-max Do you have any new comments on the PR? I didn't see your approve

ActiveChooN commented:
> But when we dump a project with one 3D tasks **it shows all the formats** (at the same time I cannot create a project with both 3D and 2D data). Is this expected behaviour?

@bsekachev, fixed

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -893,11 +904,9 @@ def _make_image(i, **kwargs):
                         attributes["track_id"] = -1
-                        index += 1
 
                 dm_item = datumaro.DatasetItem(id=osp.split(frame_data.name)[-1].split('.')[0],
```
Fix to my reported problem 3: Use either `os.path.splitext` or `str.rsplit`

zhiltsov-max commented:
My comments:

1. ![image](https://user-images.githubusercontent.com/13832349/129529446-a146e110-43a1-4668-a9c0-9c18acdddfa0.png)
  I think, it would be more user-friendly, if delete had red background and was positioned after exporting.
2. ![image](https://user-images.githubusercontent.com/13832349/129530793-da6a7c1b-b672-4924-8281-5121de38cf24.png)
  After the button is pressed and the menu disappears, when I reopen the menu, it has the button highlighted.
3. Export in sly pcd of a task with images:

```
и.м.г@ 1.JPEG
им. 536.45.--+=.г 3.jpeg
им.png.г 2.jpg
имг 3 (копия 1).jpeg
```

> Error: Request failed with status code 500. "datumaro.components.errors.RepeatedItemError: Item ('им', 'qqqaaq') is repeated in the source sequence.\n".

4. 2d project dataset export menu has 3d formats

ActiveChooN commented:
@zhiltsov-max

> I think, it would be more user-friendly, if delete had a red background and was positioned after exporting.

Fixed. I'm not sure about the red color on menu items, there is already a red button on the delete modal.

> Export in sly pcd of a task with images:

Fixed

Other comments are not very related to this PR so I moved them to a separate issue.

ActiveChooN commented:
@dvkruchinin, hi, can you add tests for this feature, please?

dvkruchinin commented:
> @dvkruchinin, hi, can you add tests for this feature, please?

Sure. A test will be prepared.
